<center>
<img src="../../img/ods_stickers.jpg" />
    
## [mlcourse.ai](mlcourse.ai) – Open Machine Learning Course 
Authors: Yury Isakov, [Yury Kashnitskiy](https://yorko.github.io) (@yorko). Edited by Anna Tarelina (@feuerengel), and Kolchenko Sergey (@KolchenkoSergey). This material is subject to the terms and conditions of the [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license. Free use is permitted for any non-commercial purpose.

# <center> Assignment #4
## <center>  User Identification with Logistic Regression (beating baselines in the "Alice" competition)

Today we are going to practice working with sparse matrices, training Logistic Regression models, and doing feature engineering. We will reproduce a couple of baselines in the ["Catch Me If You Can: Intruder Detection through Webpage Session Tracking"](https://www.kaggle.com/c/catch-me-if-you-can-intruder-detection-through-webpage-session-tracking2) (a.k.a. "Alice") Kaggle inclass competition. More credits will be given for beating a stronger baseline. 

**Your task:**
 1. "Follow me". Complete the missing code and submit your answers via [the google-form](https://docs.google.com/forms/d/1V4lHXkjZvpDDvHAcnH6RuEQJecBaLo8zooxDl1_aP60). 14 credit max. for this part
 2. "Freeride". Come up with good features to beat the baseline "A4 baseline 3". You need to name your [team](https://www.kaggle.com/c/catch-me-if-you-can-intruder-detection-through-webpage-session-tracking2/team) (out of 1 person) in full accordance with the course rating. You can think of it as a part of the assignment. 10 more credits for beating the mentioned baseline and correct team naming. 

# Part 1. Follow me

<img src='../../img/followme_alice.png' width=50%>

*image credit [@muradosmann](https://www.instagram.com/muradosmann/?hl=en)*

In [1]:
# Import libraries and set desired options
import pickle
import numpy as np
import pandas as pd
from scipy.sparse import csr_matrix, hstack
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import roc_auc_score
from sklearn.linear_model import LogisticRegression
from matplotlib import pyplot as plt
import seaborn as sns
sns.set()

##### Problem description

In this competition, we''ll analyze the sequence of websites consequently visited by a particular person and try to predict whether this person is Alice or someone else. As a metric we will use [ROC AUC](https://en.wikipedia.org/wiki/Receiver_operating_characteristic).

### 1. Data Downloading and Transformation
Register on [Kaggle](www.kaggle.com), if you have not done it before.
Go to the competition [page](https://inclass.kaggle.com/c/catch-me-if-you-can-intruder-detection-through-webpage-session-tracking2) and download the data.

First, read the training and test sets. Then we'll explore the data in hand and do a couple of simple exercises. 

In [2]:
# Read the training and test data sets, change paths if needed
train_df = pd.read_csv('/Users/aleksandr/Desktop/all/train_sessions.csv',
                       index_col='session_id')
test_df = pd.read_csv('/Users/aleksandr/Desktop/all/test_sessions.csv',
                      index_col='session_id')

# Convert time1, ..., time10 columns to datetime type
times = ['time%s' % i for i in range(1, 11)]
train_df[times] = train_df[times].apply(pd.to_datetime)
test_df[times] = test_df[times].apply(pd.to_datetime)

# Sort the data by time
train_df = train_df.sort_values(by='time1')

# Look at the first rows of the training set
train_df.head()

site1               time1  site2               time2  site3  \
session_id                                                                
21669          56 2013-01-12 08:05:57   55.0 2013-01-12 08:05:57    NaN   
54843          56 2013-01-12 08:37:23   55.0 2013-01-12 08:37:23   56.0   
77292         946 2013-01-12 08:50:13  946.0 2013-01-12 08:50:14  951.0   
114021        945 2013-01-12 08:50:17  948.0 2013-01-12 08:50:17  949.0   
146670        947 2013-01-12 08:50:20  950.0 2013-01-12 08:50:20  948.0   

                         time3  site4               time4  site5  \
session_id                                                         
21669                      NaT    NaN                 NaT    NaN   
54843      2013-01-12 09:07:07   55.0 2013-01-12 09:07:09    NaN   
77292      2013-01-12 08:50:15  946.0 2013-01-12 08:50:15  946.0   
114021     2013-01-12 08:50:18  948.0 2013-01-12 08:50:18  945.0   
146670     2013-01-12 08:50:20  947.0 2013-01-12 08:50:21  950.0   

                         time5  ...                 time6  site7  \
session_id                      ...                                
21669                      NaT  ...                   NaT    NaN   
54843                      NaT  ...                   NaT    NaN   
77292      2013-01-12 08:50:16  ...   2013-01-12 08:50:16  948.0   
114021     2013-01-12 08:50:18  ...   2013-01-12 08:50:18  947.0   
146670     2013-01-12 08:50:21  ...   2013-01-12 08:50:21  946.0   

                         time7  site8               time8  site9  \
session_id                                                         
21669                      NaT    NaN                 NaT    NaN   
54843                      NaT    NaN                 NaT    NaN   
77292      2013-01-12 08:50:16  784.0 2013-01-12 08:50:16  949.0   
114021     2013-01-12 08:50:19  945.0 2013-01-12 08:50:19  946.0   
146670     2013-01-12 08:50:21  951.0 2013-01-12 08:50:22  946.0   

                         time9 site10              time10 target  
session_id                                                        
21669                      NaT    NaN                 NaT      0  
54843                      NaT    NaN                 NaT      0  
77292      2013-01-12 08:50:17  946.0 2013-01-12 08:50:17      0  
114021     2013-01-12 08:50:19  946.0 2013-01-12 08:50:20      0  
146670     2013-01-12 08:50:22  947.0 2013-01-12 08:50:22      0  

[5 rows x 21 columns]

The training data set contains the following features:

- **site1** – id of the first visited website in the session
- **time1** – visiting time for the first website in the session
- ...
- **site10** – id of the tenth visited website in the session
- **time10** – visiting time for the tenth website in the session
- **target** – target variable, 1 for Alice's sessions, and 0 for the other users' sessions
    
User sessions are chosen in the way that they are shorter than 30 min. long and contain no more than 10 websites. I.e. a session is considered over either if a user has visited 10 websites or if a session has lasted over 30 minutes.

There are some empty values in the table, it means that some sessions contain less than ten websites. Replace empty values with 0 and change columns types to integer. Also load the websites dictionary and check how it looks like:

In [3]:
# Change site1, ..., site10 columns type to integer and fill NA-values with zeros
sites = ['site%s' % i for i in range(1, 11)]
train_df[sites] = train_df[sites].fillna(0).astype(np.uint16)
test_df[sites] = test_df[sites].fillna(0).astype(np.uint16)

# Load websites dictionary
with open(r"/Users/aleksandr/Desktop/all/site_dic.pkl", "rb") as input_file:
    site_dict = pickle.load(input_file)

# Create dataframe for the dictionary
sites_dict = pd.DataFrame(list(site_dict.keys()), index=list(site_dict.values()), columns=['site'])
print(u'Websites total:', sites_dict.shape[0])
sites_dict.head()

Websites total: 48371


site
25075              www.abmecatronique.com
13997                     groups.live.com
42436  majeureliguefootball.wordpress.com
30911           cdt46.media.tourinsoft.eu
8104                  www.hdwallpapers.eu

#### 4.1. What are the dimensions of the training and test sets (in exactly this order)? 
*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a4_q1__*

- (82797, 20) and (253561, 20)
- (82797, 20) and (253561, 21)
- (253561, 21) and (82797, 20) 
- (253561, 20) and (82797, 20)

In [5]:
train_df

site1               time1  site2               time2  site3  \
session_id                                                                
21669          56 2013-01-12 08:05:57     55 2013-01-12 08:05:57      0   
54843          56 2013-01-12 08:37:23     55 2013-01-12 08:37:23     56   
77292         946 2013-01-12 08:50:13    946 2013-01-12 08:50:14    951   
114021        945 2013-01-12 08:50:17    948 2013-01-12 08:50:17    949   
146670        947 2013-01-12 08:50:20    950 2013-01-12 08:50:20    948   
242171        952 2013-01-12 08:50:22    947 2013-01-12 08:50:23    953   
57157         953 2013-01-12 08:50:25    947 2013-01-12 08:50:26    946   
240201        946 2013-01-12 08:50:28    947 2013-01-12 08:50:28    954   
210686        946 2013-01-12 08:50:31    956 2013-01-12 08:50:32    946   
98804         948 2013-01-12 08:50:37    946 2013-01-12 08:50:37    948   
113494         52 2013-01-12 08:51:05     52 2013-01-12 08:51:27     52   
223837        513 2013-01-12 08:51:36   1116 2013-01-12 08:51:36    747   
145475       4222 2013-01-12 08:52:27   3358 2013-01-12 08:52:28   4222   
186633      38667 2013-01-12 08:53:09    181 2013-01-12 08:53:09     23   
45193          56 2013-01-12 08:53:25     55 2013-01-12 08:53:26    679   
102930        570 2013-01-12 08:55:10     23 2013-01-12 08:55:11     21   
15789         820 2013-01-12 08:59:34    820 2013-01-12 08:59:36    980   
205868         48 2013-01-12 09:10:04     56 2013-01-12 09:31:21     55   
21102        5898 2013-01-12 09:14:59   5898 2013-01-12 09:15:00   5794   
193454       2570 2013-01-12 09:15:38   2570 2013-01-12 09:15:39     21   
252924         22 2013-01-12 09:20:03     39 2013-01-12 09:20:03   1102   
12713         177 2013-01-12 09:23:33    178 2013-01-12 09:23:33    175   
182016         56 2013-01-12 09:25:03     55 2013-01-12 09:25:04    570   
55750          56 2013-01-12 09:37:32     55 2013-01-12 09:37:32      0   
237109         56 2013-01-12 09:56:03     55 2013-01-12 09:56:03     55   
120126       4277 2013-01-12 10:02:11    342 2013-01-12 10:02:11    733   
140105        733 2013-01-12 10:02:46     49 2013-01-12 10:04:02  17232   
129357        753 2013-01-12 10:05:08    532 2013-01-12 10:05:08   1020   
136699        733 2013-01-12 10:05:32    733 2013-01-12 10:05:33    733   
113189      38668 2013-01-12 10:06:19    733 2013-01-12 10:06:34    733   
...           ...                 ...    ...                 ...    ...   
131402         23 2014-04-30 16:54:28     21 2014-04-30 16:54:33    229   
72349         229 2014-04-30 16:57:08    880 2014-04-30 16:57:09     21   
181940         38 2014-04-30 16:57:26     80 2014-04-30 16:58:00     80   
67431        6003 2014-04-30 17:04:42   6003 2014-04-30 17:04:53   6004   
133256         75 2014-04-30 17:06:42    617 2014-04-30 17:06:42     77   
78516          23 2014-04-30 17:06:45     76 2014-04-30 17:06:46     80   
159245         77 2014-04-30 17:06:49     80 2014-04-30 17:06:54     76   
6103           78 2014-04-30 17:06:55     76 2014-04-30 17:06:56     77   
126232         78 2014-04-30 17:07:01     22 2014-04-30 17:07:12     22   
62543         270 2014-04-30 17:07:19    270 2014-04-30 17:07:20    270   
122173      10747 2014-04-30 17:07:36   4890 2014-04-30 17:07:36  10747   
182723         21 2014-04-30 17:10:08    229 2014-04-30 17:10:08    177   
216796         77 2014-04-30 17:10:46     77 2014-04-30 17:11:04    812   
88873          76 2014-04-30 17:11:09     80 2014-04-30 17:11:09    617   
201960         78 2014-04-30 17:11:19    881 2014-04-30 17:11:19     80   
99918          77 2014-04-30 17:14:44    812 2014-04-30 17:14:45     76   
35923          76 2014-04-30 17:15:24     81 2014-04-30 17:15:24   1307   
91326          76 2014-04-30 17:15:30    816 2014-04-30 17:15:32    812   
182878         22 2014-04-30 17:18:43     81 2014-04-30 17:20:12     77   
36912          80 2014-04-30 17:24:43     80 2014-04-30 17:24:53   2338   
101406        774 2014-04-30 17:29:53

In [6]:
test_df

site1               time1  site2               time2  site3  \
session_id                                                                
1              29 2014-10-04 11:19:53     35 2014-10-04 11:19:53     22   
2             782 2014-07-03 11:00:28    782 2014-07-03 11:00:53    782   
3              55 2014-12-05 15:55:12     55 2014-12-05 15:55:13     55   
4            1023 2014-11-04 10:03:19   1022 2014-11-04 10:03:19     50   
5             301 2014-05-16 15:05:31    301 2014-05-16 15:05:32    301   
6            1218 2014-12-05 11:21:52   1220 2014-12-05 11:21:52   1202   
7           43986 2014-05-21 14:10:43  43986 2014-05-21 14:10:44  43986   
8            3346 2014-05-25 18:32:04   3346 2014-05-25 18:33:59     23   
9            1062 2014-06-05 10:25:52     21 2014-06-05 10:25:52     23   
10             77 2014-05-02 11:48:47     77 2014-05-02 11:48:48     75   
11            301 2014-12-05 12:11:34    301 2014-12-05 12:11:35     66   
12            625 2014-12-02 08:16:47    322 2014-12-02 08:16:51    148   
13             20 2014-10-02 11:38:04     21 2014-10-02 11:38:04     21   
14            263 2014-08-04 10:02:04     21 2014-08-04 10:02:04    263   
15            722 2014-12-02 15:15:32    722 2014-12-02 15:15:43    722   
16          22889 2014-05-27 11:56:15     29 2014-05-27 11:56:15    147   
17             23 2014-05-03 16:36:18     21 2014-05-03 16:36:43     22   
18             23 2014-11-04 10:35:53    749 2014-11-04 10:35:53    747   
19             52 2014-11-02 12:15:18   1981 2014-11-02 12:15:27   3379   
20            780 2014-08-04 08:27:47    778 2014-08-04 08:27:49    778   
21             21 2014-05-03 12:20:28     35 2014-05-03 12:20:30     29   
22             23 2014-05-03 23:29:36    570 2014-05-03 23:33:38    570   
23            568 2014-05-19 14:05:32    747 2014-05-19 14:05:32     23   
24             22 2014-07-01 21:06:17   1453 2014-07-01 21:06:27     22   
25             76 2014-05-04 15:04:24   1057 2014-05-04 15:04:24   1057   
26           2891 2014-05-28 16:19:11     48 2014-05-28 16:19:15   1984   
27          30151 2014-12-02 10:55:08  30151 2014-12-02 10:55:09  30150   
28            803 2014-05-03 09:27:13    803 2014-05-03 09:27:15    803   
29             21 2014-10-02 15:10:27     21 2014-10-02 15:10:39     21   
30           1202 2014-12-05 18:13:49   1202 2014-12-05 18:14:00   1202   
...           ...                 ...    ...                 ...    ...   
82768        1202 2014-05-16 15:46:49   1202 2014-05-16 15:46:51   1202   
82769         883 2014-10-02 13:29:11    884 2014-10-02 13:29:11   5256   
82770         678 2014-05-28 17:01:59    775 2014-05-28 17:02:12    678   
82771       23695 2014-11-02 10:51:02  23695 2014-11-02 10:51:04  23695   
82772          33 2014-06-01 14:05:35   5392 2014-06-01 14:05:35     35   
82773         270 2014-06-02 15:53:01    167 2014-06-02 15:53:02    167   
82774        1102 2014-10-02 08:34:05     21 2014-10-02 08:36:02     22   
82775        3393 2014-07-03 10:47:05   3393 2014-07-03 10:47:07   3393   
82776         122 2014-05-03 09:33:35    134 2014-05-03 09:33:36    134   
82777         301 2014-05-22 10:22:08     48 2014-05-22 10:22:08     76   
82778        1086 2014-06-01 15:59:21   1086 2014-06-01 15:59:23   1086   
82779         300 2014-05-19 14:09:38    300 2014-05-19 14:09:41    302   
82780          66 2014-07-05 08:34:08     63 2014-07-05 08:34:08     21   
82781        3996 2014-07-04 17:05:53   3413 2014-07-04 17:05:53   3996   
82782          32 2014-08-02 10:11:06     30 2014-08-02 10:11:06     33   
82783         747 2014-05-25 16:43:32     38 2014-05-25 16:43:32     23   
82784        5598 2014-05-02 11:03:37   5598 2014-05-02 11:03:38   5598   
82785          22 2014-06-02 12:25:22    814 2014-06-02 12:25:22    570   
82786         786 2014-07-04 10:38:09    786 2014-07-04 10:38:13    782   
82787        6801 2014-11-02 08:44:53   6801 2014-11-02 08:45:05     22   
82788          69 2014-12-05 12:10:09

### 2. Brief Exploratory Data Analysis

In [9]:
train_df.head()

site1               time1  site2               time2  site3  \
session_id                                                                
21669          56 2013-01-12 08:05:57     55 2013-01-12 08:05:57      0   
54843          56 2013-01-12 08:37:23     55 2013-01-12 08:37:23     56   
77292         946 2013-01-12 08:50:13    946 2013-01-12 08:50:14    951   
114021        945 2013-01-12 08:50:17    948 2013-01-12 08:50:17    949   
146670        947 2013-01-12 08:50:20    950 2013-01-12 08:50:20    948   

                         time3  site4               time4  site5  \
session_id                                                         
21669                      NaT      0                 NaT      0   
54843      2013-01-12 09:07:07     55 2013-01-12 09:07:09      0   
77292      2013-01-12 08:50:15    946 2013-01-12 08:50:15    946   
114021     2013-01-12 08:50:18    948 2013-01-12 08:50:18    945   
146670     2013-01-12 08:50:20    947 2013-01-12 08:50:21    950   

                         time5  ...                 time6 site7  \
session_id                      ...                               
21669                      NaT  ...                   NaT     0   
54843                      NaT  ...                   NaT     0   
77292      2013-01-12 08:50:16  ...   2013-01-12 08:50:16   948   
114021     2013-01-12 08:50:18  ...   2013-01-12 08:50:18   947   
146670     2013-01-12 08:50:21  ...   2013-01-12 08:50:21   946   

                         time7 site8               time8 site9  \
session_id                                                       
21669                      NaT     0                 NaT     0   
54843                      NaT     0                 NaT     0   
77292      2013-01-12 08:50:16   784 2013-01-12 08:50:16   949   
114021     2013-01-12 08:50:19   945 2013-01-12 08:50:19   946   
146670     2013-01-12 08:50:21   951 2013-01-12 08:50:22   946   

                         time9 site10              time10 target  
session_id                                                        
21669                      NaT      0                 NaT      0  
54843                      NaT      0                 NaT      0  
77292      2013-01-12 08:50:17    946 2013-01-12 08:50:17      0  
114021     2013-01-12 08:50:19    946 2013-01-12 08:50:20      0  
146670     2013-01-12 08:50:22    947 2013-01-12 08:50:22      0  

[5 rows x 21 columns]

Before we start training models, we have to perform Exploratory Data Analysis ([EDA](https://en.wikipedia.org/wiki/Exploratory_data_analysis)). Today, we are going to perform a shorter version, but we will use other techniques as we move forward. Let's check which websites in the training data set are the most visited. As you can see, they are Google services and a bioinformatics website (a website with 'zero'-index is our missed values, just ignore it):

In [10]:
train_df[sites]

site1  site2  site3  site4  site5  site6  site7  site8  site9  \
session_id                                                                  
21669          56     55      0      0      0      0      0      0      0   
54843          56     55     56     55      0      0      0      0      0   
77292         946    946    951    946    946    945    948    784    949   
114021        945    948    949    948    945    946    947    945    946   
146670        947    950    948    947    950    952    946    951    946   
242171        952    947    953    946    947    946    953    955    946   
57157         953    947    946    953    955    947    953    946    953   
240201        946    947    954    953    946    954    946    956    957   
210686        946    956    946    946    955    954    946    946    946   
98804         948    946    948    784     49     53    812    982     52   
113494         52     52     52    747    747    747     23    747    568   
223837        513   1116    747     23    747    747     29     49     52   
145475       4222   3358   4222   3356   4222   3870     21   3870   3358   
186633      38667    181     23    181  38667     23    182    181  38667   
45193          56     55    679    676    814     22     39    815    752   
102930        570     23     21   3356     48    222    570    513   1379   
15789         820    820    980     49     56     55     49    784    222   
205868         48     56     55     55      0      0      0      0      0   
21102        5898   5898   5794     21   5794   5794     21    182     23   
193454       2570   2570     21     21   1102   1102     23    570    707   
252924         22     39   1102     21   1102   1102   1102     21    177   
12713         177    178    175     48     21     22    176    175    177   
182016         56     55    570    570   1038    847     22    774    570   
55750          56     55      0      0      0      0      0      0      0   
237109         56     55     55   1038     52     53     21   4277     21   
120126       4277    342    733    733    733    733    733     23    733   
140105        733     49  17232     50     48     52     23     21    989   
129357        753    532   1020    989   1021   1021     48    990    544   
136699        733    733    733    733    733    733  17232    989  17232   
113189      38668    733    733    733    733    733    568     21   1126   
...           ...    ...    ...    ...    ...    ...    ...    ...    ...   
131402         23     21    229     21    229     21    229     21    229   
72349         229    880     21     21    229    229     21     38  37261   
181940         38     80     80     80    812     76     75    879     78   
67431        6003   6003   6004      0      0      0      0      0      0   
133256         75    617     77     80     76    879     22     75     77   
78516          23     76     80    879     78     76     22     75     77   
159245         77     80     76    879     23     81     77   1307     81   
6103           78     76     77   1440     22  37239     76    617  37238   
126232         78     22     22    617     81     76     80     81     75   
62543         270    270    270    167    167     21     52     21  10747   
122173      10747   4890  10747    812     80     22     76    178    175   
182723         21    229    177    812     81     76     80     75     77   
216796         77     77    812     80     75   1307    881     76     80   
88873          76     80    617    881   1439    881     76     22     80   
201960         78    881     80     76    881     80    812     76     77   
99918          77    812     76     22     77     80     75     80     75   
35923          76     81   1307     77   1442     76  37239    617     22   
91326          76    816    812     78     76     22    617     81     80   
182878         22     81     77     80    812     76     77     77     77   
36912   

In [39]:
alisa = train_df[train_df['target'] == 1]

In [40]:
top_sites = pd.Series(alisa[sites].values.flatten()
                     ).value_counts().sort_values(ascending=False).head(5)

In [41]:
print(top_sites)
sites_dict.loc[top_sites.drop(0).index]

77    1382
80    1354
76    1307
29     897
21     857
dtype: int64


KeyError: 'labels [0] not contained in axis'

In [42]:
print(top_sites)

77    1382
80    1354
76    1307
29     897
21     857
dtype: int64


In [58]:
sites_dict.loc[top_sites.index]

site
77      i1.ytimg.com
80     s.youtube.com
76   www.youtube.com
29  www.facebook.com
21     www.google.fr

In [53]:
sites_dict

SyntaxError: invalid syntax (<ipython-input-53-279b1d7c5206>, line 1)

In [27]:
# Top websites in the training data set
top_sites = pd.Series(train_df[sites].values.flatten()
                     ).value_counts().sort_values(ascending=False).head(5)
print(top_sites)
sites_dict.loc[top_sites.drop(0).index]

21     123776
0      122730
23      87619
782     77055
22      58258
dtype: int64


site
21     www.google.fr
23    www.google.com
782   annotathon.org
22   apis.google.com

In [26]:
top_sites.head()

21     123776
0      122730
23      87619
782     77055
22      58258
dtype: int64

In [ ]:
pd.crosstab(flights_df.Month, flights_df.DayOfWeek)

##### 4.2. What kind of websites does Alice visit the most?
*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a4_q2__*

- videohostings 
- social networks
- torrent trackers
- news

In [20]:
# Your code is here

Now let us look at the timestamps and try to characterize sessions as timeframes:

In [59]:
# Create a separate dataframe where we will work with timestamps
time_df = pd.DataFrame(index=train_df.index)
time_df['target'] = train_df['target']

# Find sessions' starting and ending
time_df['min'] = train_df[times].min(axis=1)
time_df['max'] = train_df[times].max(axis=1)

# Calculate sessions' duration in seconds
time_df['seconds'] = (time_df['max'] - time_df['min']) / np.timedelta64(1, 's')

time_df.head()

target                 min                 max  seconds
session_id                                                         
21669            0 2013-01-12 08:05:57 2013-01-12 08:05:57      0.0
54843            0 2013-01-12 08:37:23 2013-01-12 09:07:09   1786.0
77292            0 2013-01-12 08:50:13 2013-01-12 08:50:17      4.0
114021           0 2013-01-12 08:50:17 2013-01-12 08:50:20      3.0
146670           0 2013-01-12 08:50:20 2013-01-12 08:50:22      2.0

In [65]:
time_df['seconds'].mean()

138.4943701909994

In [68]:
time_df[time_df['target'] == 1]['seconds'].mean()

52.29647366129734

In [69]:
time_df[time_df['target'] == 0]['seconds'].mean()

139.28237232552215

In [70]:
time_df[time_df['target'] == 0].count()

target     251264
min        251264
max        251264
seconds    251264
dtype: int64

In [71]:
time_df[time_df['target'] == 1].count()

target     2297
min        2297
max        2297
seconds    2297
dtype: int64

In [72]:
time_df.count()

target     253561
min        253561
max        253561
seconds    253561
dtype: int64

In [77]:
time_df[time_df['target'] == 1]['seconds'].max()

1763.0

In [79]:
time_df[time_df['target'] == 1]['seconds'].min()

0.0

In [80]:
time_df[time_df['target'] == 0]['seconds'].max()

1800.0

In [81]:
time_df[time_df['target'] == 0]['seconds'].min()

0.0

In [82]:

time_df['seconds'].mean()

138.4943701909994

In [84]:
import numpy as np

np.var(time_df[time_df['target'] == 0]['seconds'])

88002.95933234908

In [85]:
np.var(time_df[time_df['target'] == 1]['seconds'])

23493.421463402916

In order to perform the next task, generate descriptive statistics as you did in the first assignment.

##### 4.3. Select all correct statements:
*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a4_q3__*

- on average, Alice's session is shorter than that of other users 
- more than 1% of all sessions in the dataset belong to Alice
- minimum and maximum durations of Alice's and other users' sessions are approximately the same  
- variation about the mean session duration for all users (including Alice) is approximately the same
- less than a quarter of Alice's sessions are greater than or equal to 40 seconds  

In [22]:
# Your code is here

In order to train our first model, we need to prepare the data. First of all, exclude the target variable from the training set. Now both training and test sets have the same number of columns, therefore aggregate them into one dataframe.  Thus, all transformations will be performed simultaneously on both training and test data sets. 

On the one hand, it leads to the fact that both data sets have one feature space (you don't have to worry that you forgot to transform a feature in some data sets). On the other hand, processing time will increase. 
For the enormously large sets it might turn out that it is impossible to transform both data sets simultaneously (and sometimes you have to split your transformations into several stages only for train/test data set).
In our case, with this particular data set, we are going to perform all the transformations for the whole united dataframe at once, and before training the model or making predictions we will just take its appropriate part.

In [86]:
# Our target variable
y_train = train_df['target']

# United dataframe of the initial data 
full_df = pd.concat([train_df.drop('target', axis=1), test_df])

# Index to split the training and test data sets
idx_split = train_df.shape[0]

For the very basic model, we will use only the visited websites in the session (but we will not take into account timestamp features). The point behind this data selection is: *Alice has her favorite sites, and the more often you see these sites in the session, the higher probability that this is Alice's session, and vice versa.*

Let us prepare the data, we will take only features `site1, site2, ... , site10` from the whole dataframe. Keep in mind that the missing values are replaced with zero. Here is how the first rows of the dataframe look like:

In [87]:
# Dataframe with indices of visited websites in session
full_sites = full_df[sites]
full_sites.head()

site1  site2  site3  site4  site5  site6  site7  site8  site9  \
session_id                                                                  
21669          56     55      0      0      0      0      0      0      0   
54843          56     55     56     55      0      0      0      0      0   
77292         946    946    951    946    946    945    948    784    949   
114021        945    948    949    948    945    946    947    945    946   
146670        947    950    948    947    950    952    946    951    946   

            site10  
session_id          
21669            0  
54843            0  
77292          946  
114021         946  
146670         947

Sessions are sequences of website indices, and data in this representation is useless for machine learning method (just think, what happens if we switched all ids of all websites). 

According to our hypothesis (Alice has favorite websites), we need to transform this dataframe so each website has a corresponding feature (column) and its value is equal to number of this website visits in the session. It can be done in two lines:

In [88]:
# sequence of indices
sites_flatten = full_sites.values.flatten()

# and the matrix we are looking for 
# (make sure you understand which of the `csr_matrix` constructors is used here)
# a further toy example will help you with it
full_sites_sparse = csr_matrix(([1] * sites_flatten.shape[0],
                                sites_flatten,
                                range(0, sites_flatten.shape[0]  + 10, 10)))[:, 1:]

In [89]:
full_sites_sparse.shape

(336358, 48371)

If you understand what just happened here, then you can skip the next passage (perhaps, you can handle logistic regression too?), If not, then let us figure it out.

### Important detour #1: Sparse Matrices

Let us estimate how much memory it will require to store our data in the example above. Our united dataframe contains 336 thousand samples of 48 thousand integer features in each. It's easy to calculate the required amount of memory, roughly:

$$336K * 48K * 8 bytes = 16M * 8 bytes = 128 GB,$$

(that's the [exact](http://www.wolframalpha.com/input/?i=336358*48371*8+bytes) value). Obviously, ordinary mortals have no such volumes (strictly speaking, Python may allow you to create such a matrix, but it will not be easy to do anything with it). The interesting fact is that most of the elements of our matrix are zeros. If we count non-zero elements, then it will be about 1.8 million, i.е. slightly more than 10% of all matrix elements. Such a matrix, where most elements are zeros, is called sparse, and the ratio between the number of zero elements and the total number of elements is called the sparseness of the matrix.

For the work with such matrices you can use `scipy.sparse` library, check [documentation](https://docs.scipy.org/doc/scipy-0.18.1/reference/sparse.html) to understand what possible types of sparse matrices are, how to work with them and in which cases their usage is most effective. You can learn how they are arranged, for example, in Wikipedia [article](https://en.wikipedia.org/wiki/Sparse_matrix).
Note, that a sparse matrix contains only non-zero elements, and you can get the allocated memory size like this (significant memory savings are obvious):

In [90]:
# How much memory does a sparse matrix occupy?
print('{0} elements * {1} bytes = {2} bytes'.format(full_sites_sparse.count_nonzero(), 8, 
                                                    full_sites_sparse.count_nonzero() * 8))
# Or just like this:
print('sparse_matrix_size = {0} bytes'.format(full_sites_sparse.data.nbytes))

1866898 elements * 8 bytes = 14935184 bytes
sparse_matrix_size = 14935184 bytes


Let us explore how the matrix with the websites has been formed using a mini example. Suppose we have the following table with user sessions:

| id | site1 | site2 | site3 |
|---|---|---|---|
| 1 | 1 | 0 | 0 |
| 2 | 1 | 3 | 1 |
| 3 | 2 | 3 | 4 |

There are 3 sessions, and no more than 3 websites in each. Users visited four different sites in total (there are numbers from 1 to 4 in the table cells). And let us assume that the mapping is:

 1. vk.com
 2. habrahabr.ru 
 3. yandex.ru
 4. ods.ai

If the user has visited less than 3 websites during the session, the last few values will be zero. We want to convert the original dataframe in a way that each session has a corresponding row which shows the number of visits to each particular site. I.e. we want to transform the previous table into the following form:

| id | vk.com | habrahabr.ru | yandex.ru | ods.ai |
|---|---|---|---|---|
| 1 | 1 | 0 | 0 | 0 |
| 2 | 2 | 0 | 1 | 0 |
| 3 | 0 | 1 | 1 | 1 |


To do this, use the constructor: `csr_matrix ((data, indices, indptr))` and create a frequency table (see examples, code and comments on the links above to see how it works). Here we set all the parameters explicitly for greater clarity:

In [91]:
# data, create the list of ones, length of which equal to the number of elements in the initial dataframe (9)
# By summing the number of ones in the cell, we get the frequency,
# number of visits to a particular site per session
data = [1] * 9

# To do this, you need to correctly distribute the ones in cells
# Indices - website ids, i.e. columns of a new matrix. We will sum ones up grouping them by sessions (ids)
indices = [1, 0, 0, 1, 3, 1, 2, 3, 4]

# Indices for the division into rows (sessions)
# For example, line 0 is the elements between the indices [0; 3) - the rightmost value is not included
# Line 1 is the elements between the indices [3; 6)
# Line 2 is the elements between the indices [6; 9) 
indptr = [0, 3, 6, 9]

# Aggregate these three variables into a tuple and compose a matrix
# To display this matrix on the screen transform it into the usual "dense" matrix
csr_matrix((data, indices, indptr)).todense()

matrix([[2, 1, 0, 0, 0],
        [0, 2, 0, 1, 0],
        [0, 0, 1, 1, 1]])

In [94]:
csr_matrix((data, indices, indptr)).todense()

matrix([[2, 1, 0, 0, 0],
        [0, 2, 0, 1, 0],
        [0, 0, 1, 1, 1]])

As you might have noticed, there are not four columns in the resulting matrix (corresponding to number of different websites) but five. A zero column has been added, which indicates if the session was shorter (in our mini example we took sessions of three). This column is excessive and should be removed from the dataframe (do that yourself).

##### 4.4. What is the sparseness of the matrix in our small example?
*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a4_q4__*

- 42%
- 47%
- 50% 
- 53%




In [29]:
# Your code is here

Another benefit of using sparse matrices is that there are special implementations of both matrix operations and machine learning algorithms for them, which sometimes allows to significantly accelerate operations due to the data structure peculiarities. This applies to logistic regression as well. Now everything is ready to build our first model.

### 3. Training the first model

So, we have an algorithm and data for it. Let us build our first model, using [logistic regression](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html) implementation from ` Sklearn` with default parameters. We will use the first 90% of the data for training (the training data set is sorted by time), and the remaining 10% for validation. Let's write a simple function that returns the quality of the model and then train our first classifier:

In [95]:
def get_auc_lr_valid(X, y, C=1.0, seed=17, ratio = 0.9):
    # Split the data into the training and validation sets
    idx = int(round(X.shape[0] * ratio))
    # Classifier training
    lr = LogisticRegression(C=C, random_state=seed, solver='liblinear').fit(X[:idx, :], y[:idx])
    # Prediction for validation set
    y_pred = lr.predict_proba(X[idx:, :])[:, 1]
    # Calculate the quality
    score = roc_auc_score(y[idx:], y_pred)
    
    return score

In [96]:
%%time
# Select the training set from the united dataframe (where we have the answers)
X_train = full_sites_sparse[:idx_split, :]

# Calculate metric on the validation set
print(get_auc_lr_valid(X_train, y_train))

0.9195245587148385
CPU times: user 7.82 s, sys: 346 ms, total: 8.17 s
Wall time: 6.25 s


The first model demonstrated the quality  of 0.92 on the validation set. Let's take it as the first baseline and starting point. To make a prediction on the test data set **we need to train the model again on the entire training data set** (until this moment, our model used only part of the data for training), which will increase its generalizing ability:

In [97]:
# Function for writing predictions to a file
def write_to_submission_file(predicted_labels, out_file,
                             target='target', index_label="session_id"):
    predicted_df = pd.DataFrame(predicted_labels,
                                index = np.arange(1, predicted_labels.shape[0] + 1),
                                columns=[target])
    predicted_df.to_csv(out_file, index_label=index_label)

In [98]:
# Train the model on the whole training data set
# Use random_state=17 for repeatability
# Parameter C=1 by default, but here we set it explicitly
lr = LogisticRegression(C=1.0, random_state=17, solver='liblinear').fit(X_train, y_train)

# Make a prediction for test data set
X_test = full_sites_sparse[idx_split:,:]
y_test = lr.predict_proba(X_test)[:, 1]

# Write it to the file which could be submitted
write_to_submission_file(y_test, 'baseline_1.csv')

If you follow these steps and upload the answer to the competition [page](https://inclass.kaggle.com/c/catch-me-if-you-can-intruder-detection-through-webpage-session-tracking2), you will get `ROC AUC = 0.90812` on the public leaderboard ("A4 baseline 1").

### 4. Model Improvement: Feature Engineering

Now we are going to try to improve the quality of our model by adding new features to the data. But first, answer the following question:

##### 4.5. What years are present in the training and test datasets, if united?
*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a4_q5__*

- 13 and 14
- 2012 and 2013
- 2013 and 2014 
- 2014 and 2015

In [101]:
full_df.head()

site1               time1  site2               time2  site3  \
session_id                                                                
21669          56 2013-01-12 08:05:57     55 2013-01-12 08:05:57      0   
54843          56 2013-01-12 08:37:23     55 2013-01-12 08:37:23     56   
77292         946 2013-01-12 08:50:13    946 2013-01-12 08:50:14    951   
114021        945 2013-01-12 08:50:17    948 2013-01-12 08:50:17    949   
146670        947 2013-01-12 08:50:20    950 2013-01-12 08:50:20    948   

                         time3  site4               time4  site5  \
session_id                                                         
21669                      NaT      0                 NaT      0   
54843      2013-01-12 09:07:07     55 2013-01-12 09:07:09      0   
77292      2013-01-12 08:50:15    946 2013-01-12 08:50:15    946   
114021     2013-01-12 08:50:18    948 2013-01-12 08:50:18    945   
146670     2013-01-12 08:50:20    947 2013-01-12 08:50:21    950   

                         time5  site6               time6  site7  \
session_id                                                         
21669                      NaT      0                 NaT      0   
54843                      NaT      0                 NaT      0   
77292      2013-01-12 08:50:16    945 2013-01-12 08:50:16    948   
114021     2013-01-12 08:50:18    946 2013-01-12 08:50:18    947   
146670     2013-01-12 08:50:21    952 2013-01-12 08:50:21    946   

                         time7  site8               time8  site9  \
session_id                                                         
21669                      NaT      0                 NaT      0   
54843                      NaT      0                 NaT      0   
77292      2013-01-12 08:50:16    784 2013-01-12 08:50:16    949   
114021     2013-01-12 08:50:19    945 2013-01-12 08:50:19    946   
146670     2013-01-12 08:50:21    951 2013-01-12 08:50:22    946   

                         time9  site10              time10  
session_id                                                  
21669                      NaT       0                 NaT  
54843                      NaT       0                 NaT  
77292      2013-01-12 08:50:17     946 2013-01-12 08:50:17  
114021     2013-01-12 08:50:19     946 2013-01-12 08:50:20  
146670     2013-01-12 08:50:22     947 2013-01-12 08:50:22

Create a feature that will be a number in YYYYMM format from the date when the session was held, for example 201407 -- year 2014 and 7th month. Thus, we will take into account the monthly [linear trend](http://people.duke.edu/~rnau/411trend.htm) for the entire period of the data provided.

In [102]:
# Dataframe for new features
full_new_feat = pd.DataFrame(index=full_df.index)

# Add start_month feature
full_new_feat['start_month'] = full_df['time1'].apply(lambda ts: 
                                                      100 * ts.year + ts.month).astype('float64')

In [105]:
full_new_feat.head()

start_month
session_id             
21669          201301.0
54843          201301.0
77292          201301.0
114021         201301.0
146670         201301.0

In [118]:
full_new_feat['start_month']

session_id
21669     201301.0
54843     201301.0
77292     201301.0
114021    201301.0
146670    201301.0
242171    201301.0
57157     201301.0
240201    201301.0
210686    201301.0
98804     201301.0
113494    201301.0
223837    201301.0
145475    201301.0
186633    201301.0
45193     201301.0
102930    201301.0
15789     201301.0
205868    201301.0
21102     201301.0
193454    201301.0
252924    201301.0
12713     201301.0
182016    201301.0
55750     201301.0
237109    201301.0
120126    201301.0
140105    201301.0
129357    201301.0
136699    201301.0
113189    201301.0
            ...   
82768     201405.0
82769     201410.0
82770     201405.0
82771     201411.0
82772     201406.0
82773     201406.0
82774     201410.0
82775     201407.0
82776     201405.0
82777     201405.0
82778     201406.0
82779     201405.0
82780     201407.0
82781     201407.0
82782     201408.0
82783     201405.0
82784     201405.0
82785     201406.0
82786     201407.0
82787     201411.0
82788     201412.0
8

In [114]:
full_new_feat.groupby['start_month'].count

TypeError: 'method' object is not subscriptable

##### 4.6. Plot the graph of the number of Alice sessions versus the new feature, start_month. Choose the correct statement:

*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a4_q6__*

- Alice wasn't online at all for the entire period
- From the beginning of 2013 to mid-2014, the number of Alice's sessions per month decreased
- The number of Alice's sessions per month is generally constant for the entire period
- From the beginning of 2013 to mid-2014, the number of Alice's sessions per month increased

*Hint: the graph will be more explicit if you treat `start_month` as a categorical ordinal variable*.

\frac{\frac{n}{2} + \frac{m}{2}}{m+n}

In [107]:
import matplotlib.pyplot as plt; plt.rcdefaults()
import numpy as np
import matplotlib.pyplot as plt

In this way, we have an illustration and thoughts about the usefulness of the new feature, add it to the training sample and check the quality of the new model:

In [115]:
# Add the new feature to the sparse matrix
tmp = full_new_feat[['start_month']].values
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], tmp[:idx_split,:]]))

# Compute the metric on the validation set
print(get_auc_lr_valid(X_train, y_train))

0.7508354860175162


The quality of the model has decreased significantly. We added a feature that definitely seemed useful to us, but its usage only worsened the model. Why did it happen?

### Important detour #2: is it necessary to scale features?

Here we give an intuitive reasoning (a rigorous mathematical justification for one or another aspect in linear models you can easily find on the internet). Consider the features more closely: those of them that correspond to the number of visits to a particular web-site per session vary from 0 to 10. The feature `start_month` has a completely different range: from 201301 to 201412, this means the contribution of this variable is significantly greater than the others. It would seem that problem can be avoided if we put less weight in a linear combination of attributes in this case, but in our case logistic regression with regularization is used (by default, this parameter is `C = 1`), which penalizes the model the stronger the greater its weights are. Therefore, for linear methods with regularization, it is recommended to convert features to the same scale (you can read more about the regularization, for example, [here](https://habrahabr.ru/company/ods/blog/322076/)).

One way to do this is standardization: for each observation you need to subtract the average value of the feature and divide this difference by the standard deviation:

$$ x^{*}_{i} = \dfrac{x_{i} - \mu_x}{\sigma_x}$$

The following practical tips can be given:
- It is recommended to scale features if they have essentially different ranges or different units of measurement (for example, the country's population is indicated in units, and the country's GNP in trillions)
- Scale features if you do not have a reason/expert opinion to give a greater weight to any of them
- Scaling can be excessive if the ranges of some of your features differ from each other, but they are in the same system of units (for example, the proportion of middle-aged people and people over 80 among the entire population)
- If you want to get an interpreted model, then build a model without regularization and scaling (most likely, its quality will be worse)
- Binary features (which take only values of 0 or 1) are usually left without conversion, (but)
- If the quality of the model is crucial, try different options and select one where the quality is better

Getting back to `start_month`, let us rescale the new feature and train the model again. This time the quality has increased:

In [116]:
# Add the new standardized feature to the sparse matrix
tmp = StandardScaler().fit_transform(full_new_feat[['start_month']])
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], tmp[:idx_split,:]]))

# Compute metric on the validation set
print(get_auc_lr_valid(X_train, y_train))

0.9196830663159573


In [119]:
tmp

array([[-2.07907592],
       [-2.07907592],
       [-2.07907592],
       ...,
       [ 0.55213183],
       [ 0.55213183],
       [ 0.70393227]])

In [117]:
train_df.head()

site1               time1  site2               time2  site3  \
session_id                                                                
21669          56 2013-01-12 08:05:57     55 2013-01-12 08:05:57      0   
54843          56 2013-01-12 08:37:23     55 2013-01-12 08:37:23     56   
77292         946 2013-01-12 08:50:13    946 2013-01-12 08:50:14    951   
114021        945 2013-01-12 08:50:17    948 2013-01-12 08:50:17    949   
146670        947 2013-01-12 08:50:20    950 2013-01-12 08:50:20    948   

                         time3  site4               time4  site5  \
session_id                                                         
21669                      NaT      0                 NaT      0   
54843      2013-01-12 09:07:07     55 2013-01-12 09:07:09      0   
77292      2013-01-12 08:50:15    946 2013-01-12 08:50:15    946   
114021     2013-01-12 08:50:18    948 2013-01-12 08:50:18    945   
146670     2013-01-12 08:50:20    947 2013-01-12 08:50:21    950   

                         time5  ...                 time6 site7  \
session_id                      ...                               
21669                      NaT  ...                   NaT     0   
54843                      NaT  ...                   NaT     0   
77292      2013-01-12 08:50:16  ...   2013-01-12 08:50:16   948   
114021     2013-01-12 08:50:18  ...   2013-01-12 08:50:18   947   
146670     2013-01-12 08:50:21  ...   2013-01-12 08:50:21   946   

                         time7 site8               time8 site9  \
session_id                                                       
21669                      NaT     0                 NaT     0   
54843                      NaT     0                 NaT     0   
77292      2013-01-12 08:50:16   784 2013-01-12 08:50:16   949   
114021     2013-01-12 08:50:19   945 2013-01-12 08:50:19   946   
146670     2013-01-12 08:50:21   951 2013-01-12 08:50:22   946   

                         time9 site10              time10 target  
session_id                                                        
21669                      NaT      0                 NaT      0  
54843                      NaT      0                 NaT      0  
77292      2013-01-12 08:50:17    946 2013-01-12 08:50:17      0  
114021     2013-01-12 08:50:19    946 2013-01-12 08:50:20      0  
146670     2013-01-12 08:50:22    947 2013-01-12 08:50:22      0  

[5 rows x 21 columns]

##### 4.7. Add to the training set a new feature "n_unique_sites" – the number of the unique web-sites in a session. Calculate how the quality on the validation set has changed

*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a4_q7__*

- It has decreased. It is better not to add a new feature. 
- It has not changed
- It has decreased. The new feature should be scaled.
- I am confused, and I do not know if it's necessary to scale a new feature.

*Tips: use the nunique() function from `pandas`. Do not forget to include the start_month in the set. Will you scale a new feature? Why?*

In [126]:
new = train_df[sites].nunique()

In [127]:
# Add the new standardized feature to the sparse matrix
tmp = StandardScaler().fit_transform(new)
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], tmp[:idx_split,:]]))

# Compute metric on the validation set
print(get_auc_lr_valid(X_train, y_train))

ValueError: Expected 2D array, got 1D array instead:
array=[15765. 15822. 15760. 15776. 15841. 15769. 15773. 15855. 15725. 15676.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

In [121]:
train_df['start_month'] = full_new_feat['start_month']

ValueError: cannot reindex from a duplicate axis

In [39]:
# Your code is here

So, the new feature has slightly decreased the quality, so we will not use it. Nevertheless, do not rush to throw features out because they haven't performed well. They can be useful in a combination with other features (for example, when a new feature is a ratio or a product of two others).

#####  4.8. Add two new features: start_hour and morning. Calculate the metric. Which of these features gives an improvement?

The `start_hour` feature is the hour at which the session started (from 0 to 23), and the binary feature `morning` is equal to 1 if the session started in the morning and 0 if the session started later (we assume that morning means `start_hour` is equal to 11 or less).

Will you scale the new features? Make your assumptions and test them in practice.

*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a4_q8__*

- None of the features gave an improvement :(
- `start_hour` feature gave an improvement, and `morning` did not
- `morning` feature gave an improvement, and `start_hour` did not
- Both features gave an improvement

*Tip: find suitable functions for working with time series data in [documentation](http://pandas.pydata.org/pandas-docs/stable/api.html). Do not forget to include the `start_month` feature.*

In [40]:
# Your code is here

### 5. Regularization and Parameter Tuning

We have introduced features that improve the quality of our model in comparison with the first baseline. Can we do even better? After we have changed the training and test sets, it almost always makes sense to search for the optimal hyperparameters - the parameters of the model that do not change during training.

For example, in week 3, you learned that, in decision trees, the depth of the tree is a hyperparameter, but the feature by which splitting occurs and its threshold is not. 

In the logistic regression that we use, the weights of each feature are changing, and we find their optimal values during training; meanwhile, the regularization parameter remains constant. This is the hyperparameter that we are going to optimize now.

Calculate the quality on a validation set with a regularization parameter, which is equal to 1 by default:

In [122]:
# Compose the training set
tmp_scaled = StandardScaler().fit_transform(full_new_feat[['start_month', 
                                                           'start_hour', 
                                                           'morning']])
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], 
                             tmp_scaled[:idx_split,:]]))

# Capture the quality with default parameters
score_C_1 = get_auc_lr_valid(X_train, y_train)
print(score_C_1)

KeyError: "['start_hour' 'morning'] not in index"

We will try to beat this result by optimizing the regularization parameter. We will take a list of possible values of C and calculate the quality metric on the validation set for each of C-values:

In [ ]:
from tqdm import tqdm

# List of possible C-values
Cs = np.logspace(-3, 1, 10)
scores = []
for C in tqdm(Cs):
    scores.append(get_auc_lr_valid(X_train, y_train, C=C))

Plot the graph of the quality metric (AUC-ROC) versus the value of the regularization parameter. The value of quality metric corresponding to the default value of C=1 is represented by a horizontal dotted line:

In [ ]:
plt.plot(Cs, scores, 'ro-')
plt.xscale('log')
plt.xlabel('C')
plt.ylabel('AUC-ROC')
plt.title('Regularization Parameter Tuning')
# horizontal line -- model quality with default C value
plt.axhline(y=score_C_1, linewidth=.5, color='b', linestyle='dashed') 
plt.show()

##### 4.9. What is the value of parameter C (if rounded to 2 decimals) that corresponds to the highest model quality?

*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel #mlcourse_ai, pinned thread __#a4_q9__*

- 0.17 
- 0.46
- 1.29
- 3.14

In [ ]:
# Your code is here

For the last task in this assignment: train the model using the optimal regularization parameter you found (do not round up to two digits like in the last question). If you do everything correctly and submit your solution, you should see `ROC AUC = 0.92784` on the public leaderboard ("A4 baseline 2"):

In [ ]:
# Prepare the training and test data
tmp_scaled = StandardScaler().fit_transform(full_new_feat[['start_month', 'start_hour', 
                                                           'morning']])
X_train = csr_matrix(hstack([full_sites_sparse[:idx_split,:], 
                             tmp_scaled[:idx_split,:]]))
X_test = csr_matrix(hstack([full_sites_sparse[idx_split:,:], 
                            tmp_scaled[idx_split:,:]]))

# Train the model on the whole training data set using optimal regularization parameter
lr = LogisticRegression(C=C, random_state=17, solver='liblinear').fit(X_train, y_train)

# Make a prediction for the test set
y_test = lr.predict_proba(X_test)[:, 1]

# Write it to the submission file
write_to_submission_file(y_test, 'baseline_2.csv')

In this part of the assignment, you have learned how to use sparse matrices, train logistic regression models, create new features and selected the best ones, learned why you need to scale features, and how to select hyperparameters. That's a lot!

# Part 2. Freeride

<img src='../../img/snowboard.jpg' width=70%>

*Yorko in Sheregesh, the best palce in Russia for snowboarding and skiing.*

In this part, you'll need to beat the "A4 baseline 3" baseline. No more step-by-step instructions. But it'll be very helpful for you to study the Kernel "[Correct time-aware cross-validation scheme](https://www.kaggle.com/kashnitsky/correct-time-aware-cross-validation-scheme)".

Here are a few tips for finding new features: think about what you can come up with using existing features, try multiplying or dividing two of them, justify or decline your hypotheses with plots, extract useful information from time series data (time1 ... time10), do not hesitate to convert an existing feature (for example, take a logarithm), etc. Checkout other [Kernels](https://www.kaggle.com/c/catch-me-if-you-can-intruder-detection-through-webpage-session-tracking2/kernels). We encourage you to try new ideas and models throughout the course and participate in the competitions - it's fun!

When you get into Kaggle and Xgboost, you'll feel like that, and it's OK :)

<img src='../../img/xgboost_meme.jpg' width=50%>